experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+		runtime: IFluidDatastoreRuntime,
+		id?: string,
+		params?: Partial<SharedSomethingOptions>,
+	): SharedSomething {
+		// Uh oh--params unused!
+		return runtime.createChannel(id, SharedSomethingFactory.Type) as SharedSomething;
+	}
+
+	public static getFactory(params?: Partial<SharedSomethingOptions>): SharedSomethingFactory {
+		return new SharedSomethingFactory(params);
+	}
+}
+```
+
+This immediately uncovers one issue: since the static `create` method isn't in charge of constructing the DDS's factory,
+there's no place to plumb through the parameters.
+Even worse, the container author likely already wrote some code like this:
+
+```typescript
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	SharedSomething.getFactory({ formatVersion: 5 }),
+]);
+```
+
+If they later write `SharedSomething.create(runtime, { formatVersion: 4 })`, the create flow outlined above will instead use the
+factory created at data object registry construction time.
+
+There is another, more subtle issue with this strategy: how should one implement a `SharedSomethingFactory` which takes in some parameters?
+
+```typescript
+class SharedSomethingFactory {
+	public constructor(private readonly options: SharedSomethingOptions) {}
+
+	public load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		channelAttributes: Readonly<IChannelAttributes>,
+	): Promise<IChannel> {
+		const sharedSomething = new SharedSomething(
+			runtime,
+			id,
+			this.options /* uh oh: channelAttributes and this.options might disagree on snapshot format! */,
+		);
+		await sharedSomething.load(services);
+		return sharedSomething;
+	}
+
+	public create(runtime: IFluidDataStoreRuntime, id: string): IChannel {
+		const sharedSomething = new SharedSomething(document, id, this.options);
+		sharedSomething.initializeLocal();
+		return sharedSomething;
+	}
+}
+```
+
+If the channel's attributes clearly reflect its parameters (e.g. the snapshot format being used),
+then the resulting factory sometimes produces DDSes which don't match the requested format!
+
+There are essentially two ways to resolve this issue:
+
+1. Make `SharedSomethingFactory` have a parameterless constructor, and pass parameters all the way through the `create` flow (i.e. through `createChannel`, factory lookup, and factory's `create` method).
+2. Keep parameters in `SharedSomethingFactory`'s constructor, but change their semantics such that they only necessarily apply to newly created documents.
+
+Regardless of which option we pick, enough information to reify the DDS's parameters needs to be persisted in the attributes blob.
+
+In order to support data objects which can create multiple types of the same shared object with different parameters, option 2 would need to be accompanied by changes to make DDS registration take in named entries in `@fluidframework/aqueduct`, rather than assume the "key" for a DDS registry entry is that DDS's factory's `type` field.
```

Comment:
IMO we should take named entries everywhere anyway.  No need for factories to know their own name.

Reply:
Also, the blocker for this isn't aqueduct so much as that our SharedObject factories currently stamp their own  attributes in on creation (they have no way to consult the name they were registered under).

Reply:
Probably FluidDataStoreContext should be stamping its own "registryName" into the summary in FluidDataStoreContext.summarizeInternal, rather than adopting whatever the data store thought its "type" was as the registryName.  Haven't looked at it too hard though.